src/lib.rs

```diff
@@ -12,34 +12,59 @@
 /// Modules for handling Input/Output of GeoTIFF data
 pub mod io;
 
-use std::fs::File;
+use std::io::Cursor;
 
 use ndarray::Dim;
 use numpy::{PyArray, ToPyArray};
+use object_store::{parse_url, ObjectStore};
 use pyo3::prelude::{pyfunction, pymodule, PyModule, PyResult, Python};
 use pyo3::wrap_pyfunction;
+use tokio;
+use url::Url;
 
 /// Read a GeoTIFF file from a path on disk into an ndarray
 ///
 /// Parameters
 /// ----------
 /// path : str
-///     The path to the file.
+///     The path to the file, or a url to a remote file.
 ///
 /// Returns
 /// -------
 /// array : np.ndarray
 ///     2D array containing the GeoTIFF pixel data.
+///
+/// Examples
+/// --------
+/// from cog3pio import read_geotiff
+///
+/// array = read_geotiff("https://github.com/pka/georaster/raw/v0.1.0/data/tiff/float32.tif")
+/// assert array.shape == (20, 20)
 #[pyfunction]
 #[pyo3(name = "read_geotiff")]
 fn read_geotiff_py<'py>(
     path: &str,
     py: Python<'py>,
 ) -> PyResult<&'py PyArray<f32, Dim<[usize; 2]>>> {
-    // Open TIFF file from path
-    let file = File::open(path).expect("Cannot find GeoTIFF file");
+    // Initialize async runtime
+    let runtime = tokio::runtime::Builder::new_current_thread()
+        .enable_all()
+        .build()?;
+
+    // Get TIFF file stream asynchronously
+    let stream = runtime.block_on(async {
+        // Parse URL
+        let url = Url::parse(path).expect(&format!("Cannot parse path: {path}"));
+        let (store, location) = parse_url(&url).expect(&format!("Cannot parse url: {url}"));
+
+        // Return cursor to in-memory buffer
+        let result = store.get(&location).await.unwrap();
```

Comment:
Need to find a way here to let `object_store` read from local filepaths like `/tmp/float32.tif` or `data/float32.tif`. Currently at commit 33faafb3f43ac48aaee62a0a4c999774815ab768, one would need to prepend the filepath with `file://` for the read to work, otherwise there will be an error like `pyo3_runtime.PanicException: Cannot parse path: /tmp/tmp6ip79h4_/float32.tif: RelativeUrlWithoutBase` (see https://github.com/weiji14/cog3pio/actions/runs/8115987634/job/22185036928?pr=5#step:6:89).

Maybe add some logic to determine the URL scheme, and if it is local, prepend the url with `file://`?

Reply:
Oh wait, [`RelativeUrlWithoutBase`](https://docs.rs/url/2.5.0/url/enum.ParseError.html#variant.RelativeUrlWithoutBase) is raised on `Url::parse`, not ObjectStore :sweat_smile:

Reply:
Great, need to figure out how to handle Windows paths - https://github.com/weiji14/cog3pio/actions/runs/8118215740/job/22192052940?pr=5#step:6:95:

```
E       pyo3_runtime.PanicException: Cannot parse url: c:\Users\RUNNER~1\AppData\Local\Temp\tmpv5aci1cm\float32.tif: Generic { store: "URL", source: Unrecognised { url: Url { scheme: "c", cannot_be_a_base: true, username: "", password: None, host: None, port: None, path: "\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmpv5aci1cm\\float32.tif", query: None, fragment: None } } }
```

Reply:
Ok, figured out that there's a [`Url::from_file_path`](https://docs.rs/url/2.5.0/url/struct.Url.html#method.from_file_path) function that handles Windows paths (and also prepends `file://` to local filepaths). Implemented in commit 3c8a604e306dd103335e0b4e3263ad7ee6817a4d